### ISSUE TITLE
DRAFT: Plotting protocol

### ISSUE THREAD

**Issue by henryiii** [scikit-hep/hist#64]
I am working on an API for plotting libraries, so that they can read and plot from any Histogram library without depending on that library. Here is roughly what I'm thinking:

```python
from numpy.typing import ArrayLike # requires NumPy master
from typing import Protocol, Optional, Tuple, Union, Iterable]

class PlottableAxis(Protocol):
    # In this version, name and title are not linked by the histogram library
    # It is a library's job to plot name if title is blank
    name: str
    title: str

    edges: ArrayLike

     # For non-categorical axes, this returns None
    categories: Union[Iterable[int], Iterable[str], None]

class PlottableHistogram(Protocol):
    axes : Tuple[PlottableAxis]
   
    # Values returns the array or the values array for specialized accumulators
    def values(self, flow=False) -> ArrayLike: ...

    # Variance returns the variance if applicable, otherwise None
    def variances(self, flow=False) -> Optional[ArrayLike]: ...
```

Current thoughts:
* Should we use `axes[i].edges` and something like `.view()`? Note that `.view()` will not work since it does not always return a plottable array. Maybe we could have `.values` and `.errors: Optional`, which would work for all storages? Errors would return None if no errors present?
    - values returns the array or the 'values' array for specialized accumulors
    - variances returns the variances if applicable for that Storage, otherwise None
* Should the name if title not set logic be in the plotting library or in the the underlying object?
* Plotting libraries should be able to access flow bins if they need to - the current form does have a way for them to get the flow, but not for the axes to indicate how they align. `to_numpy` would support this.

**Comment by henryiii** [comment]
This is assuming the master object is the API. You could also have a `__histogram_dict__` that returns something like what you see above - then it's much easier to add to existing libraries and doesn't affect/change the public API, but on the flip side, it's duplicated and an extra requirement. (Double underscores for clarity - you really aren't supposed to add anything with double underscores, so for an API choice, it might not be ideal).

**Comment by henryiii** [comment]
Released in boost-histogram 0.13, hist 2.1.0, uproot 4, and mplhep 0.2.16!
